server/service/src/asset/insert.rs

```diff
@@ -0,0 +1,112 @@
+use super::{query::get_asset, validate::check_asset_exists};
+use crate::{service_provider::ServiceContext, SingleRecordError};
+use chrono::{NaiveDate, Utc};
+use repository::{
+    assets::{
+        asset::{AssetFilter, AssetRepository},
+        asset_row::{AssetRow, AssetRowRepository},
+    },
+    RepositoryError, StorageConnection, StringFilter,
+};
+
+#[derive(PartialEq, Debug)]
+pub enum InsertAssetError {
+    AssetAlreadyExists,
+    CreatedRecordNotFound,
+    DatabaseError(RepositoryError),
+    SerialNumberAlreadyExists,
+}
+
+pub struct InsertAsset {
+    pub id: String,
+    pub store_id: Option<String>,
+    pub name: String,
+    pub code: String,
+    pub serial_number: Option<String>,
+    pub catalogue_item_id: Option<String>,
+    pub installation_date: Option<NaiveDate>,
+    pub replacement_date: Option<NaiveDate>,
+}
+
+pub fn insert_asset(
+    ctx: &ServiceContext,
+    input: InsertAsset,
+) -> Result<AssetRow, InsertAssetError> {
+    let asset = ctx
+        .connection
+        .transaction_sync(|connection| {
+            validate(&input, connection)?;
+            let new_asset = generate(input);
+            AssetRowRepository::new(&connection).upsert_one(&new_asset)?;
+
+            get_asset(ctx, new_asset.id).map_err(InsertAssetError::from)
+        })
+        .map_err(|error| error.to_inner_error())?;
+    Ok(asset)
+}
+
+pub fn validate(
+    input: &InsertAsset,
+    connection: &StorageConnection,
+) -> Result<(), InsertAssetError> {
+    if check_asset_exists(&input.id, connection)?.is_some() {
+        return Err(InsertAssetError::AssetAlreadyExists);
+    }
+
+    // Check the serial number is unique (if present)
+    if let Some(serial_number) = &input.serial_number {
+        if AssetRepository::new(connection)
+            .query_one(AssetFilter::new().serial_number(StringFilter::equal_to(serial_number)))?
+            .is_some()
+        {
+            return Err(InsertAssetError::SerialNumberAlreadyExists);
+        }
+    }
+
+    // TODO: Do we require Name, and or Code? I'm assuming user can leave them as empty strings if they want?
+
+    Ok(())
+}
+
+pub fn generate(
+    InsertAsset {
+        id,
+        store_id,
+        name,
+        code,
+        serial_number,
+        catalogue_item_id,
+        installation_date,
+        replacement_date,
+    }: InsertAsset,
+) -> AssetRow {
+    AssetRow {
+        id,
+        store_id,
+        name,
+        code,
+        serial_number,
+        catalogue_item_id,
+        installation_date,
+        replacement_date,
+        created_datetime: Utc::now().naive_utc(),
+        modified_datetime: Utc::now().naive_utc(),
+        deleted_datetime: None,
+    }
+}
```

Comment:
I think asset_category_id, asset_type_id, and asset_catalogue_item_id are missing from AssetRow type and this generate function (or somewhere else in service layer?)

I get foreign key violation errors when I attempt to insert any asset.

Reply:
I've removed them from `AssetRow` entirely after a discussion with Mark Prins, however I see there's still included in the migration I'll fix that. The idea is that we just need the `catalogue_item_id` link and we can look up that other related data.
You may need to re-initialise you database to get rid of those columns, or delete them manually though...